shell/platform/android/io/flutter/embedding/engine/systemchannels/RestorationChannel.java

```diff
@@ -0,0 +1,144 @@
+ * getRestorationData()}.
+ */
+public class RestorationChannel {
+  private static final String TAG = "RestorationChannel";
+
+  public RestorationChannel(
+      @NonNull DartExecutor dartExecutor, @NonNull boolean waitForRestorationData) {
+    this(
+        new MethodChannel(dartExecutor, "flutter/restoration", StandardMethodCodec.INSTANCE),
+        waitForRestorationData);
+  }
+
+  RestorationChannel(MethodChannel channel, @NonNull boolean waitForRestorationData) {
+    this.channel = channel;
+    this.waitForRestorationData = waitForRestorationData;
+
+    channel.setMethodCallHandler(handler);
+  }
+
+  /**
+   * Whether {@code setRestorationData} will be called to provide restoration data for the
+   * framework.
+   *
+   * <p>When this is set to true, the channel will delay answering any requests for restoration data
+   * by the framework until {@code setRestorationData} has been called. It must be set to false if
+   * the engine never calls {@code setRestorationData}. If it has been set to true, but it later
+   * turns out that there is no restoration data, {@code setRestorationData} must be called with
+   * null.
+   */
+  public final boolean waitForRestorationData;
+
+  private MethodChannel channel;
+  private byte[] restorationData;
+  private MethodChannel.Result pendingResult;
+  private boolean engineHasProvidedData = false;
+  private boolean frameworkHasRequestedData = false;
+
+  /** Obtain the most current restoration data that the framework has provided. */
+  public byte[] getRestorationData() {
+    return restorationData;
+  }
+
+  /** Set the restoration data from which the framework will restore its state. */
+  public void setRestorationData(byte[] data) {
+    engineHasProvidedData = true;
+    if (pendingResult != null) {
+      pendingResult.success(data);
+      pendingResult = null;
+      restorationData = data;
+    } else if (frameworkHasRequestedData) {
+      channel.invokeMethod(
+          "push",
+          data,
+          new MethodChannel.Result() {
+            @Override
+            public void success(Object result) {
+              restorationData = data;
+            }
+
+            @Override
+            public void error(String errorCode, String errorMessage, Object errorDetails) {
+              Log.e(
+                  TAG,
+                  "Error "
+                      + errorCode
+                      + " while sending restoration data to framework: "
+                      + errorMessage);
+            }
+
+            @Override
+            public void notImplemented() {
+              // Nothing to do.
+            }
+          });
+    } else {
+      restorationData = data;
+    }
+  }
+
+  /**
+   * Clears the current restoration data.
+   *
+   * <p>This should be called just prior to a hot restart. Otherwise, after the hot restart the
+   * state prior to the hot restart will get restored.
+   */
+  public void clearData() {
```

Comment:
Why not do it more frequently? There's some 2 directional complexity here but I still think it should be possible to find logical situations where this data is deterministically not useful no?

Reply:
> If we clear it on one side, we'd also need to clear it on the other

what does this mean specifically? There's only memory in one place no?

Reply:
The restoration state is duplicated on the engine and framework side because of the threading issue: The operating system will ask for the restoration state on the platform thread, the state itself is owned by the framework running on the UI thread, though. When the operating system asks for it, we can't reach into the UI thread to retrieve it (see https://docs.google.com/document/d/1KIiq5CdqnSXxQXbZIDy2Ukc-JHFyLak1JR8e2cm3eO4/edit#heading=h.xhk980xrzj8j). Therefore, we must maintain an up-to-date copy of the restoration state in the engine. The two copies are kept in sync via this channel: When the operating system provides restoration state to the engine, it sends it over the framework. When the restoration state in the framework changes, it's send back to the engine (so it can be given to the operating system when it asks for it again).

When we clear the state on one side without telling the other, we break that contract and I don't see a good reason to do it. The only place where we could delete the engine copy is after we've handed it to the operating system. In reality, the OS only asks for it moments before it kills the entire app anyway (and thus frees up the memory). So why introduce the extra complexity of sometimes engine and framework restoration state can be out of sync?

Reply:
discussed offline. SG